aten/src/ATen/native/cpu/MultinomialKernel.cpp

```diff
@@ -36,85 +37,189 @@ void multinomial_apply(Tensor& result, const Tensor& self, const int64_t n_sampl
 
+      prev_sum = sum;
       sum += val;
+      if ((prev_sum == sum && val != 0) || (sum == val && prev_sum != 0)){
+        // We tried to summarize very large and small numbers and lost precision.
+        // We will use extra memory that will hold integer and fractional part of the sum separately.
+        lost_precision = true;
+        break;
+      }
       if (val == 0) {
         n_zeros += 1;
       }
       cum_dist_ptr[j * cum_dist_stride_0] = sum;
     }
 
-    TORCH_CHECK(sum > 0, "invalid multinomial distribution (sum of probabilities <= 0)");
-    TORCH_CHECK(with_replacement || (n_categories - n_zeros >= n_sample),
-        "invalid multinomial distribution (with replacement=False, not enough non-negative category to sample)");
+    if (!lost_precision) {
+      TORCH_CHECK(sum > 0, "invalid multinomial distribution (sum of probabilities <= 0)");
+      TORCH_CHECK(with_replacement || (n_categories - n_zeros >= n_sample),
+                  "invalid multinomial distribution (with replacement=False, not enough non-negative category to sample)");
 
-    /* normalize cumulative probability distribution so that last val is 1
-    i.e. doesn't assume original self row sums to one */
-    if ((sum > 0) || ((sum < 1.00001) && (sum > 0.99999))) {
-      for (int64_t j = 0; j < n_categories; j++) {
-        cum_dist_ptr[j * cum_dist_stride_0] /= sum;
+      /* normalize cumulative probability distribution so that last val is 1
+      i.e. doesn't assume original self row sums to one */
+      if ((sum > 0) || ((sum < 1.00001) && (sum > 0.99999))) {
+        for (int64_t j = 0; j < n_categories; j++) {
+          cum_dist_ptr[j * cum_dist_stride_0] /= sum;
+        }
       }
-    }
 
-    for (int64_t j = 0; j < n_sample; j++) {
-      /* sample a probability mass from a uniform distribution */
-      at::uniform_real_distribution<double> uniform(0, 1);
-      double uniform_sample = uniform(gen);
-      /* Do a binary search for the slot in which the prob falls
-      ie cum_dist[row][slot-1] < uniform_prob < cum_distr[row][slot] */
-      int left_pointer = 0;
-      int right_pointer = n_categories;
-      int mid_pointer;
-      scalar_t cum_prob;
-      int sample_idx;
-      /* Make sure the last cumulative distribution bucket sums to 1 */
-      cum_dist_ptr[(n_categories - 1) * cum_dist_stride_0] = 1;
-
-      while(right_pointer - left_pointer > 0) {
-        mid_pointer = left_pointer + (right_pointer - left_pointer) / 2;
-        cum_prob = cum_dist_ptr[mid_pointer * cum_dist_stride_0];
-        if (cum_prob < uniform_sample) {
-          left_pointer = mid_pointer + 1;
-        }
-        else {
-          right_pointer = mid_pointer;
+      for (int64_t j = 0; j < n_sample; j++) {
+        /* sample a probability mass from a uniform distribution */
+        at::uniform_real_distribution<double> uniform(0, 1);
+        double uniform_sample = uniform(gen);
+        /* Do a binary search for the slot in which the prob falls
```

Comment:
BTW, it would be nice to use std::lower_bound so you don't have to reimplement binary search